src/core/xep/carbons.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2017 Stephen Paul Weber
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "module.h"
+#include "misc.h"
+#include "settings.h"
+#include "signals.h"
+
+#include "rosters-tools.h"
+#include "xmpp-servers.h"
+#include "disco.h"
+#include "tools.h"
+#include "muc.h"
+
+#define XMLNS_CARBONS "urn:xmpp:carbons:2"
+
+static void
+sig_server_features(XMPP_SERVER_REC *server)
+{
+	if (disco_have_feature(server->server_features, XMLNS_CARBONS)) {
+		LmMessage *lmsg;
+		LmMessageNode *node;
+
+		lmsg = lm_message_new_with_sub_type(
+			NULL,
+			LM_MESSAGE_TYPE_IQ,
+			LM_MESSAGE_SUB_TYPE_SET
+		);
+		node = lm_message_node_add_child(lmsg->node, "enable", NULL);
+		lm_message_node_set_attribute(node, XMLNS, XMLNS_CARBONS);
+		signal_emit("xmpp send iq", 2, server, lmsg);
+		lm_message_unref(lmsg);
+	}
+}
+
+static void
+sig_recv_message(XMPP_SERVER_REC *server, LmMessage *lmsg, const int type,
+    const char *id, const char *from, const char *to)
+{
+	LmMessageNode *node = NULL;
+	LmMessageNode *forwarded = NULL;
+	LmMessageNode *message = NULL;
+	char *ffrom, *fto;
+
+	node = lm_find_node(lmsg->node, "received", "xmlns", XMLNS_CARBONS);
+	if(node) {
```

Comment:
```suggestion
	if (node != NULL) {
```
same with the other ifs.